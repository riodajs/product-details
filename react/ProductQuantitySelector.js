import ProductQuantitySelector from './components/ProductQuantitySelector/index'
export default ProductQuantitySelector
