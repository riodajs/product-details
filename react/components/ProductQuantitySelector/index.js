import React, { useCallback } from 'react'
import PropTypes from 'prop-types'
import { NumericStepper } from 'vtex.styleguide'
import { FormattedMessage } from 'react-intl'
import { ProductContext } from 'vtex.product-context'
import { path, isEmpty } from 'ramda'

import styles from './quantitySelector.css'

const ProductQuantitySelector = ({ warningQuantityThreshold, ...props }) => {
  const valuesFromContext = React.useContext(ProductContext)

  const numericStepperProps = () => {
    if (!valuesFromContext || isEmpty(valuesFromContext)) {
      const { onChange, availableQuantity, selectedQuantity } = props

      return {
        availableQuantity,
        selectedQuantity,
        onChange: useCallback(e => onChange(e.values), []),
      }
    }

    const { selectedItem, selectedQuantity, onChangeQuantity } = valuesFromContext
    return {
      availableQuantity: path(['sellers', 0, 'commertialOffer', 'AvailableQuantity'], selectedItem),
      selectedQuantity,
      onChange: useCallback(e => onChangeQuantity(e.value), []),
    }
  }

  const { availableQuantity, onChange, selectedQuantity } = numericStepperProps()
  const showAvailable = availableQuantity <= warningQuantityThreshold

  if (availableQuantity < 1) return null

  return (
    <div className={`${styles.quantitySelectorContainer} flex flex-column mb4`}>
      <div className="mb3 c-muted-2 t-body">
        <FormattedMessage id="product-details.quantity" />
      </div>
      <NumericStepper
        size="small"
        value={selectedQuantity}
        minValue={1}
        maxValue={availableQuantity ? availableQuantity : undefined}
        onChange={onChange}
      />
      {showAvailable && 
        <div className={`${styles.availableQuantityContainer} mv4 c-muted-2 t-small`}>
          <FormattedMessage id="product-details.quantity-available" values={{ availableQuantity }} />
        </div>}
    </div>
  )
}

ProductQuantitySelector.defaultProps = {
  warningQuantityThreshold: 0,
  onChange: () => {},
}

ProductQuantitySelector.propTypes = {
  selectedQuantity: PropTypes.number.isRequired,
  availableQuantity: PropTypes.number,
  onChange: PropTypes.func,
  warningQuantityThreshold: PropTypes.number.isRequired
}

ProductQuantitySelector.schema = {
  title: 'editor.product-quantity-selector.title',
  description: 'editor.product-quantity-selector.description',
  type: 'object',
  properties: {
    warningQuantityThreshold: {
      title: 'editor.product-quantity-selector.warningQuantityThreshold.title',
      description: 'editor.product-quantity-selector.warningQuantityThreshold.title',
      type: 'number',
      default: 0,
    },
  },
}

export default ProductQuantitySelector