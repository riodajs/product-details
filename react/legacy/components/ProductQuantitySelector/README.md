# VTEX Product Quantity Selector

## Description

The VTEX Product Quantity Selector allows the user to add to cart the specified amount of the displayed product

## Table of Contents

- [Usage](#usage)
  - [Blocks API](#blocks-api)
    - [Configuration](#configuration)
  - [Styles API](#styles-api)
    - [CSS Namespaces](#css-namespaces)

## Usage

This app uses our store builder with the blocks architecture. To know more about Store Builder [click here.](https://help.vtex.com/en/tutorial/understanding-storebuilder-and-stylesbuilder#structuring-and-configuring-our-store-with-object-object)

To use this app or override the default CSS you need import it in your dependencies on `manifest.json` file.

```json
  "dependencies": {
    "vtex.product-details": "1.x"
  }
```

This block may be used inside the `product-details` block.

To use it, you must declare its use in your `product-details` block array. An example of usage in a `blocks.json`:

```json
"product-details#default": {
    "blocks": [
      "breadcrumb",
      "product-name",
      "product-images",
      "product-price",
      "product-description",
      "product-specifications",
      "buy-button",
      "sku-selector",
      "shipping-simulator",
      "availability-subscriber",
      "share",
      "product-quantity-selector"
    ],
    "props": {
      "displayVertically": true,
      "share": {
        "social": {
          "Facebook": true,
          "WhatsApp": true,
          "Twitter": false
        }
      },
      "price": {
        "labelSellingPrice": null,
        "showListPrice": true,
        "showLabels": true,
        "showInstallments": true,
        "showSavings": true
      },
      "name": {
        "showBrandName": false,
        "showSku": false,
        "showProductReference": false
      }
    }
  },
"product-quantity-selector": {
    "props": {
      "warningQuantityThreshold": 9999999
    }
  },
```

### Blocks API

This is the defined interface for this block:

```json
{
  "product-quantity-selector": {
    "component": "ProductQuantitySelectorLegacy"
  }
}
```

For now this block does not have any required or optional blocks.

#### Configuration

Through the Storefront, you can change the Product Quantity Selector's behavior and interface. However, you also can make in your theme app, as Store theme does.

| Prop name           | Type      | Description                                                                                 |
| ------------------- | --------- | ------------------------------------------------------------------------------------------- |
| `warningQuantityThreshold`     | `Number` | Only show the quantity of remaining items in stock if item available quantity is less than or equal to the value passed in this property. Default: 0 (does not appear).    |

### Styles API

This app provides some CSS classes as an API for style customization.

To use this CSS API, you must add the `styles` builder and create an app styling CSS file.

1. Add the `styles` builder to your `manifest.json`:

```json
  "builders": {
    "styles": "1.x"
  }
```

2. Create a file called `vtex.product-details.css` inside the `styles/css` folder. Add your custom styles:

```css
.quantitySelectorContainer {
  margin-top: 10px;
}
```

#### CSS Namespaces

Below, we describe the namespaces that are defined in the Product Quantity Selector.

| Token name                 | Component                                                                                                                                                                                                                                                                                                                                                                     | Description                                                   |
| -------------------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ------------------------------------------------------------- |
| `quantitySelectorContainer`                | [index](https://github.com/vtex-apps/product-details/blob/master/react/components/ProductQuantitySelector/index.js) | The main container of `Product Quantity Selector`                        |
| `availableQuantityContainer`                | [index](https://github.com/vtex-apps/product-details/blob/master/react/components/ProductQuantitySelector/index.js) | The container that wraps the available quantity view.                        |