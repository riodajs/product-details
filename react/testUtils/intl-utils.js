import React from 'react'
import { render } from 'react-testing-library'
import { IntlProvider } from 'react-intl'
import defautMessages from '../../messages/en.json'

export const renderWithIntl = (node, options) => {
  const rendered = render(
    <IntlProvider messages={defautMessages} locale="en-US">
      {node}
    </IntlProvider>,
    options
  )

  return {
    ...rendered,
    rerender: newUi =>
      renderWithIntl(newUi, {
        container: rendered.container,
        baseElement: rendered.baseElement,
      }),
  }
}
