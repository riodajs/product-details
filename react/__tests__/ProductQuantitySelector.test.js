import React from 'react'
import { renderWithIntl } from 'intl-utils'

import ProductQuantitySelector from '../ProductQuantitySelector'

describe('<ProductQuantitySelector /> component', () => {
  it('should render and not show available quantity view', () => {
    const component = renderWithIntl(<ProductQuantitySelector selectedQuantity={1} availableQuantity={50} maximumAvailableQuantity={10}/>)
    expect(component).toBeDefined()
    expect(component.asFragment()).toMatchSnapshot()
  })
  it('should render and show available quantity view', () => {
    const component = renderWithIntl(<ProductQuantitySelector selectedQuantity={1} availableQuantity={5} maximumAvailableQuantity={10}/>)
    expect(component).toBeDefined()
    expect(component.asFragment()).toMatchSnapshot()
  })
})