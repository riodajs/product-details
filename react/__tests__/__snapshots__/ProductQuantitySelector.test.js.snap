// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductQuantitySelector /> component should render and not show available quantity view 1`] = `
<DocumentFragment>
  <div
    class="quantitySelectorContainer flex flex-column mb4"
  >
    <div
      class="mb3 c-muted-2 t-body"
    >
      <span>
        Quantity
      </span>
    </div>
    NumericStepper props={"size":"small","value":1,"minValue":1,"maxValue":50}
  </div>
</DocumentFragment>
`;

exports[`<ProductQuantitySelector /> component should render and show available quantity view 1`] = `
<DocumentFragment>
  <div
    class="quantitySelectorContainer flex flex-column mb4"
  >
    <div
      class="mb3 c-muted-2 t-body"
    >
      <span>
        Quantity
      </span>
    </div>
    NumericStepper props={"size":"small","value":1,"minValue":1,"maxValue":5}
    <div
      class="mv4 c-muted-2 t-small"
    >
      <span>
        5 available
      </span>
    </div>
  </div>
</DocumentFragment>
`;
